src/compare.cc

```diff
@@ -0,0 +1,158 @@
+#include <node.h>
+#include <vips/vips.h>
+
+#include "nan.h"
+
+#include "common.h"
+#include "compare-internal.h"
+
+using v8::Boolean;
+using v8::Exception;
+using v8::Function;
+using v8::Handle;
+using v8::Local;
+using v8::Number;
+using v8::Object;
+using v8::String;
+using v8::Value;
+
+using sharp::ImageType;
+using sharp::DetermineImageType;
+using sharp::InitImage;
+using sharp::counterQueue;
+
+struct CompareBaton {
+  // Input
+  std::string fileIn1;
+  std::string fileIn2;
+
+  // Output
+  float meanSquaredError;
+  std::string err;
+
+  CompareBaton():
+    meanSquaredError(0.0) {}
+};
+
+class CompareWorker : public NanAsyncWorker {
+
+ public:
+  CompareWorker(NanCallback *callback, CompareBaton *baton) : NanAsyncWorker(callback), baton(baton) {}
+  ~CompareWorker() {}
+
+  void Execute() {
+    // Decrement queued task counter
+    g_atomic_int_dec_and_test(&counterQueue);
+
+    ImageType imageType1 = ImageType::UNKNOWN;
+    ImageType imageType2 = ImageType::UNKNOWN;
+    VipsImage *image1 = NULL;
+    VipsImage *image2 = NULL;
+
+    // Create "hook" VipsObject to hang image references from
+    hook = reinterpret_cast<VipsObject*>(vips_image_new());
+
+    // From files
+    imageType1 = DetermineImageType(baton->fileIn1.c_str());
+    imageType2 = DetermineImageType(baton->fileIn2.c_str());
+
+    if (imageType1 != ImageType::UNKNOWN) {
+      image1 = InitImage(baton->fileIn1.c_str(), VIPS_ACCESS_RANDOM);
+      if (image1 == NULL) {
+        (baton->err).append("Input file 1 has corrupt header");
+        imageType1 = ImageType::UNKNOWN;
+      }
+    } else {
+      (baton->err).append("Input file 1 is of an unsupported image format");
+    }
+
+    if (imageType2 != ImageType::UNKNOWN) {
+      image2 = InitImage(baton->fileIn2.c_str(), VIPS_ACCESS_RANDOM);
+      if (image2 == NULL) {
+        (baton->err).append("Input file 2 has corrupt header");
+        imageType2 = ImageType::UNKNOWN;
+      }
+    } else {
+      (baton->err).append("Input file 2 is of an unsupported image format");
+    }
+
+    if (image1 != NULL && imageType1 != ImageType::UNKNOWN && image2 != NULL && imageType2 != ImageType::UNKNOWN) {
+      double meanSquaredError;
+      if (Compare(hook, image1, image2, &meanSquaredError)) {
+        (baton->err).append("Failed to compare two images");
+        return Error();
+      }
+      baton->meanSquaredError = meanSquaredError;
+      g_object_unref(image1);
```

Comment:
I think you'll need to use `vips_object_local` to "hang" `image1` and `image2` on the `hook`, then use a final `g_object_unref(hook);` in the clean-up section to correctly unref everything.